File: Beecrowd_Parte1/Controlador_de_Voo_3163.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

class Aviao {
    String id;

    Aviao(String id) {
        this.id = id;
    }
}

class Fila {
    Queue<Aviao> queue;

    Fila() {
        queue = new LinkedList<>();
    }

    void adicionarNaFila(String id) {
        queue.add(new Aviao(id));
    }

    String removerDaFila() {
        return queue.poll().id;
    }

    boolean isEmpty() {
        return queue.isEmpty();
    }
}

public class Controlador_de_Voo_3163 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        Fila oeste = new Fila();
        Fila norte = new Fila();
        Fila sul = new Fila();
        Fila leste = new Fila();

        while (true) {
            int pontoCardeal = scanner.nextInt();
            if (pontoCardeal == 0) {
                break;
            }

            scanner.nextLine();  // Consumir a nova linha
            String id = scanner.nextLine();

            switch (pontoCardeal) {
                case -1:
                    oeste.adicionarNaFila(id);
                    break;
                case -2:
                    sul.adicionarNaFila(id);
                    break;
                case -3:
                    norte.adicionarNaFila(id);
                    break;
                case -4:
                    leste.adicionarNaFila(id);
                    break;
                default:
                    break;
            }
        }

        StringBuilder resultado = new StringBuilder();
        while (!oeste.isEmpty() || !norte.isEmpty() || !sul.isEmpty() || !leste.isEmpty()) {
            if (!oeste.isEmpty()) {
                resultado.append(oeste.removerDaFila()).append(" ");
            }
            if (!norte.isEmpty()) {
                resultado.append(norte.removerDaFila()).append(" ");
            }
            if (!sul.isEmpty()) {
                resultado.append(sul.removerDaFila()).append(" ");
            }
            if (!leste.isEmpty()) {
                resultado.append(leste.removerDaFila()).append(" ");
            }
        }

        System.out.println(resultado.toString().trim());

        scanner.close();
    }
}

File: Beecrowd_Parte1/Controlador_de_Voo_3163.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct Aviao {
    char id[5];
    struct Aviao* next;
} Aviao;

typedef struct {
    Aviao* head;
    Aviao* tail;
} Fila;

void inicializarFila(Fila* fila) {
    fila->head = NULL;
    fila->tail = NULL;
}

void adicionarNaFila(Fila* fila, char* id) {
    Aviao* novoAviao = (Aviao*)malloc(sizeof(Aviao));
    strcpy(novoAviao->id, id);
    novoAviao->next = NULL;

    if (fila->tail != NULL) {
        fila->tail->next = novoAviao;
    }
    fila->tail = novoAviao;

    if (fila->head == NULL) {
        fila->head = novoAviao;
    }
}

char* removerDaFila(Fila* fila) {
    if (fila->head == NULL) {
        return NULL;
    }
    Aviao* aviao = fila->head;
    fila->head = aviao->next;
    if (fila->head == NULL) {
        fila->tail = NULL;
    }
    char* id = (char*)malloc(5 * sizeof(char));
    strcpy(id, aviao->id);
    free(aviao);
    return id;
}

int main() {
    Fila oeste, norte, sul, leste;
    inicializarFila(&oeste);
    inicializarFila(&norte);
    inicializarFila(&sul);
    inicializarFila(&leste);

    int pontoCardeal;
    char id[5];

    while (1) {
        scanf("%d", &pontoCardeal);
        if (pontoCardeal == 0) {
            break;
        }

        scanf("%s", id);

        switch (pontoCardeal) {
            case -1:
                adicionarNaFila(&oeste, id);
                break;
            case -2:
                adicionarNaFila(&sul, id);
                break;
            case -3:
                adicionarNaFila(&norte, id);
                break;
            case -4:
                adicionarNaFila(&leste, id);
                break;
            default:
                break;
        }
    }

    while (oeste.head != NULL || norte.head != NULL || sul.head != NULL || leste.head != NULL) {
        if (oeste.head != NULL) {
            printf("%s ", removerDaFila(&oeste));
        }
        if (norte.head != NULL) {
            printf("%s ", removerDaFila(&norte));
        }
        if (sul.head != NULL) {
            printf("%s ", removerDaFila(&sul));
        }
        if (leste.head != NULL) {
            printf("%s ", removerDaFila(&leste));
        }
    }
    printf("\n");

    return 0;
}
